docs/architecture/ADR-003-Non-interactive-defaults.md

```diff
@@ -0,0 +1,698 @@
+
+```go
+func (p *parsedTx) malleate(txConf client.TxConfig, squareSize uint64) error {
+   if p.msg == nil || p.tx == nil {
+       return errors.New("can only malleate a tx with a MsgWirePayForData")
+   }
+
+   // parse wire message and create a single message
+   _, unsignedPFD, sig, err := types.ProcessWirePayForData(p.msg, squareSize)
+   if err != nil {
+       return err
+   }
+
+   // create the signed PayForData using the fees, gas limit, and sequence from
+   // the original transaction, along with the appropriate signature.
+   signedTx, err := types.BuildPayForDataTxFromWireTx(p.tx, txConf.NewTxBuilder(), sig, unsignedPFD)
+   if err != nil {
+       return err
+   }
+
+   rawProcessedTx, err := txConf.TxEncoder()(signedTx)
+   if err != nil {
+       return err
+   }
+
+   p.malleatedTx = rawProcessedTx
+   return nil
+}
+```
+
+When doing this process over all of the transactions, we also need to add the share index as metadata to the malleated transaction when we wrap it. This is completed in the malleateTxs step.
+
+```go
+
+func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+   ...
+   // in this step we are processing any MsgWirePayForData transactions into
+   // MsgPayForData and their respective messages. The malleatedTxs contain the
+   // the new sdk.Msg with the original tx's metadata (sequence number, gas
+   // price etc).
+   processedTxs, messages, err := malleateTxs(app.txConfig, squareSize, parsedTxs, req.BlockData.Evidence)
+   if err != nil {
+       panic(err)
+   }
+   ...
+```
+
+#### Encoding the square
+
+As briefly discussed earlier, one major change in how we are producing blocks is that we are using the normal mechanism for splitting shares.
+
+```go
+func (app *App) PrepareProposal(req abci.RequestPrepareProposal) abci.ResponsePrepareProposal {
+   ...
+   blockData := core.Data{
+       Txs:                processedTxs,
+       Evidence:           req.BlockData.Evidence,
+       Messages:           core.Messages{MessagesList: messages},
+       OriginalSquareSize: squareSize,
+   }
+
+   dataSquare, err := shares.Split(coreData)
+   if err != nil {
+       panic(err)
+   }
+   ...
+}
+```
+
+### ProcessProposal
+
+Fortunately, most of the work necessary for non-interactive defaults is encapsulated by `PrepareProposal`. Our goal in `ProcessProposal` is to enforce the constraints that we set during `PrepareProposal`. Note that we cannot actually check the last two constraints, so we don't.
+
+- All messages must be ordered lexicographically by namespace.
+- The commitments signed over in each `MsgPayForData` must consist only of subtree roots of the data square.
+- If a `MsgPayForData` is added to the square, then its corresponding message must also be included.
+- There must not be a message without a `MsgPayForData` (does this need to be a rule? cc @adlerjohn).
```

Comment:
@adlerjohn can you reply here? We should not ask questions to each other in ADRs IMO.

Reply:
Responded in my review above.